web/packages/api/src/toEthereum_v2.ts

```diff
@@ -347,38 +374,84 @@ export async function getDeliveryFee(
-        "0x0000000000000000000000000000000000000000",
-        "0x0000000000000000000000000000000000000000",
-        "0x0000000000000000000000000000000000000000000000000000000000000000",
-        340282366920938463463374607431768211455n,
-        340282366920938463463374607431768211455n,
-        340282366920938463463374607431768211455n,
-        parachain,
-        340282366920938463463374607431768211455n
-    )
+    const { tokenErcMetadata, sourceParachain, ahAssetMetadata, sourceAssetMetadata } =
+        resolveInputs(registry, tokenAddress, parachain)
 
-    let assetHubExecutionFeeDOT = 0n
-    let returnToSenderExecutionFeeDOT = 0n
-    let returnToSenderDeliveryFeeDOT = 0n
-    const bridgeHubDeliveryFeeDOT = await calculateDeliveryFee(
-        assetHub,
-        registry.bridgeHubParaId,
-        xcm
-    )
-    if (parachain !== registry.assetHubParaId) {
-        const returnToSenderXcm = buildParachainERC20ReceivedXcmOnDestination(
+    let xcm: any
+
+    if (sourceAssetMetadata.location) {
+        xcm = buildResultXcmAssetHubPNATransferFromParachain(
             assetHub.registry,
             registry.ethChainId,
+            sourceAssetMetadata,
+            "0x0000000000000000000000000000000000000000000000000000000000000000",
             "0x0000000000000000000000000000000000000000",
+            "0x0000000000000000000000000000000000000000000000000000000000000000",
             340282366920938463463374607431768211455n,
             340282366920938463463374607431768211455n,
+            340282366920938463463374607431768211455n,
+            parachain,
+            340282366920938463463374607431768211455n
+        )
+    } else {
+        xcm = buildResultXcmAssetHubERC20TransferFromParachain(
+            assetHub.registry,
+            registry.ethChainId,
             "0x0000000000000000000000000000000000000000000000000000000000000000",
-            "0x0000000000000000000000000000000000000000000000000000000000000000"
+            "0x0000000000000000000000000000000000000000",
+            "0x0000000000000000000000000000000000000000",
+            "0x0000000000000000000000000000000000000000000000000000000000000000",
+            340282366920938463463374607431768211455n,
+            340282366920938463463374607431768211455n,
+            340282366920938463463374607431768211455n,
+            parachain,
+            340282366920938463463374607431768211455n
+        )
+    }
+    console.log("xcm on asset hub for pna transfer from 3rd parachain:", xcm.toHuman())
+
+    let assetHubExecutionFeeDOT = 0n
+    let returnToSenderExecutionFeeDOT = 0n
+    let returnToSenderDeliveryFeeDOT = 0n
+    let bridgeHubDeliveryFeeDOT =
+        registry.parachains[registry.assetHubParaId].estimatedDeliveryFeeDOT || 1_000_000_000n
+    const ahParachain = registry.parachains[registry.assetHubParaId]
+    if (ahParachain.features.hasXcmPaymentApi) {
+        // Todo: Seems incorrect--the XCM here is the one executed on AH,
```

Comment:
You are correct, the original version was incorrect. So basically now we are over-estimating the size of the blob by measuring  a larger blob and in effect charging the user a slightly higher delivery fee.

Reply:
https://github.com/Snowfork/snowbridge/pull/1419/commits/8b70d62602591da474babccf8fe2ab9f64a4f026